### DISCUSSION TITLE
How to enforce separation of concerns between workspace packages?

### DISCUSSION THREAD

**Discussion by adam-arthur** [yarnpkg/berry#2356] [Ideas]
Hello!

I'm working on a project where I have multiple packages in the same workspace. I want to enforce it such that Package A can only use items exposed via Package B's "main" file. Basically, limited to the api that somebody who does an npm/yarn install would see.  

Why?
Separation of concerns. I don't want my codebase to devolve into a spaghetti-mess by having every project import implementation details from every other package. It's better to have a hard barrier, rather than rely on discipline... especially because in many cases there are files with the same name between projects.

Example:
- game
  - physics - should not be able to access render files
  - render - should not be able to access physics files
  - core - Accesses export API of physics and render to create game loop. Cannot access implementation details beyond that

I'm not sure where the best place to enforce this is though. Is it better that this be a TypeScript plugin? Should I be using Lerna? Regardless, having this capability would be a huge win for code quality and the value proposition of a monorepo setup.

**Comment by arcanis** [comment]
It'll be via the [`exports` field](https://nodejs.org/api/packages.html#packages_exports), but we don't support it just yet. Should come with Yarn 3, in a few months.

**Reply by merceyz** [reply]
Support for that is tracked in https://github.com/yarnpkg/berry/issues/650
